
<!doctype html>
<html lang="en" class="no-js">
  <head>
    
      <meta charset="utf-8">
      <meta name="viewport" content="width=device-width,initial-scale=1">
      
      
      
        <link rel="canonical" href="https://YAOJ-bioin.github.io/Plant_spatial_multiomics_analyis/1_preprocess/">
      
      
        <link rel="prev" href="..">
      
      
      
      <link rel="icon" href="../assets/images/favicon.png">
      <meta name="generator" content="mkdocs-1.6.1, mkdocs-material-9.5.50">
    
    
      
        <title>Preprocessing - Plant Spatial Multiomics Analysis</title>
      
    
    
      <link rel="stylesheet" href="../assets/stylesheets/main.a40c8224.min.css">
      
      


    
    
      
    
    
      
        
        
        <link rel="preconnect" href="https://fonts.gstatic.com" crossorigin>
        <link rel="stylesheet" href="https://fonts.googleapis.com/css?family=Roboto:300,300i,400,400i,700,700i%7CRoboto+Mono:400,400i,700,700i&display=fallback">
        <style>:root{--md-text-font:"Roboto";--md-code-font:"Roboto Mono"}</style>
      
    
    
    <script>__md_scope=new URL("..",location),__md_hash=e=>[...e].reduce(((e,_)=>(e<<5)-e+_.charCodeAt(0)),0),__md_get=(e,_=localStorage,t=__md_scope)=>JSON.parse(_.getItem(t.pathname+"."+e)),__md_set=(e,_,t=localStorage,a=__md_scope)=>{try{t.setItem(a.pathname+"."+e,JSON.stringify(_))}catch(e){}}</script>
    
      

    
    
    
  </head>
  
  
    <body dir="ltr">
  
    
    <input class="md-toggle" data-md-toggle="drawer" type="checkbox" id="__drawer" autocomplete="off">
    <input class="md-toggle" data-md-toggle="search" type="checkbox" id="__search" autocomplete="off">
    <label class="md-overlay" for="__drawer"></label>
    <div data-md-component="skip">
      
        
        <a href="#preprocess" class="md-skip">
          Skip to content
        </a>
      
    </div>
    <div data-md-component="announce">
      
    </div>
    
    
      

  

<header class="md-header md-header--shadow" data-md-component="header">
  <nav class="md-header__inner md-grid" aria-label="Header">
    <a href=".." title="Plant Spatial Multiomics Analysis" class="md-header__button md-logo" aria-label="Plant Spatial Multiomics Analysis" data-md-component="logo">
      
  
  <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M12 8a3 3 0 0 0 3-3 3 3 0 0 0-3-3 3 3 0 0 0-3 3 3 3 0 0 0 3 3m0 3.54C9.64 9.35 6.5 8 3 8v11c3.5 0 6.64 1.35 9 3.54 2.36-2.19 5.5-3.54 9-3.54V8c-3.5 0-6.64 1.35-9 3.54"/></svg>

    </a>
    <label class="md-header__button md-icon" for="__drawer">
      
      <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M3 6h18v2H3zm0 5h18v2H3zm0 5h18v2H3z"/></svg>
    </label>
    <div class="md-header__title" data-md-component="header-title">
      <div class="md-header__ellipsis">
        <div class="md-header__topic">
          <span class="md-ellipsis">
            Plant Spatial Multiomics Analysis
          </span>
        </div>
        <div class="md-header__topic" data-md-component="header-topic">
          <span class="md-ellipsis">
            
              Preprocessing
            
          </span>
        </div>
      </div>
    </div>
    
    
      <script>var palette=__md_get("__palette");if(palette&&palette.color){if("(prefers-color-scheme)"===palette.color.media){var media=matchMedia("(prefers-color-scheme: light)"),input=document.querySelector(media.matches?"[data-md-color-media='(prefers-color-scheme: light)']":"[data-md-color-media='(prefers-color-scheme: dark)']");palette.color.media=input.getAttribute("data-md-color-media"),palette.color.scheme=input.getAttribute("data-md-color-scheme"),palette.color.primary=input.getAttribute("data-md-color-primary"),palette.color.accent=input.getAttribute("data-md-color-accent")}for(var[key,value]of Object.entries(palette.color))document.body.setAttribute("data-md-color-"+key,value)}</script>
    
    
    
      <label class="md-header__button md-icon" for="__search">
        
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.52 6.52 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5"/></svg>
      </label>
      <div class="md-search" data-md-component="search" role="dialog">
  <label class="md-search__overlay" for="__search"></label>
  <div class="md-search__inner" role="search">
    <form class="md-search__form" name="search">
      <input type="text" class="md-search__input" name="query" aria-label="Search" placeholder="Search" autocapitalize="off" autocorrect="off" autocomplete="off" spellcheck="false" data-md-component="search-query" required>
      <label class="md-search__icon md-icon" for="__search">
        
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.52 6.52 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5"/></svg>
        
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M20 11v2H8l5.5 5.5-1.42 1.42L4.16 12l7.92-7.92L13.5 5.5 8 11z"/></svg>
      </label>
      <nav class="md-search__options" aria-label="Search">
        
        <button type="reset" class="md-search__icon md-icon" title="Clear" aria-label="Clear" tabindex="-1">
          
          <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M19 6.41 17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z"/></svg>
        </button>
      </nav>
      
    </form>
    <div class="md-search__output">
      <div class="md-search__scrollwrap" tabindex="0" data-md-scrollfix>
        <div class="md-search-result" data-md-component="search-result">
          <div class="md-search-result__meta">
            Initializing search
          </div>
          <ol class="md-search-result__list" role="presentation"></ol>
        </div>
      </div>
    </div>
  </div>
</div>
    
    
      <div class="md-header__source">
        <a href="https://github.com/YAOJ-bioin/Plant_spatial_multiomics_analyis" title="Go to repository" class="md-source" data-md-component="source">
  <div class="md-source__icon md-icon">
    
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 448 512"><!--! Font Awesome Free 6.7.2 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2024 Fonticons, Inc.--><path d="M439.55 236.05 244 40.45a28.87 28.87 0 0 0-40.81 0l-40.66 40.63 51.52 51.52c27.06-9.14 52.68 16.77 43.39 43.68l49.66 49.66c34.23-11.8 61.18 31 35.47 56.69-26.49 26.49-70.21-2.87-56-37.34L240.22 199v121.85c25.3 12.54 22.26 41.85 9.08 55a34.34 34.34 0 0 1-48.55 0c-17.57-17.6-11.07-46.91 11.25-56v-123c-20.8-8.51-24.6-30.74-18.64-45L142.57 101 8.45 235.14a28.86 28.86 0 0 0 0 40.81l195.61 195.6a28.86 28.86 0 0 0 40.8 0l194.69-194.69a28.86 28.86 0 0 0 0-40.81"/></svg>
  </div>
  <div class="md-source__repository">
    YAOJ-bioin/Plant_spatial_multiomics_analyis
  </div>
</a>
      </div>
    
  </nav>
  
</header>
    
    <div class="md-container" data-md-component="container">
      
      
        
          
        
      
      <main class="md-main" data-md-component="main">
        <div class="md-main__inner md-grid">
          
            
              
              <div class="md-sidebar md-sidebar--primary" data-md-component="sidebar" data-md-type="navigation" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    



<nav class="md-nav md-nav--primary" aria-label="Navigation" data-md-level="0">
  <label class="md-nav__title" for="__drawer">
    <a href=".." title="Plant Spatial Multiomics Analysis" class="md-nav__button md-logo" aria-label="Plant Spatial Multiomics Analysis" data-md-component="logo">
      
  
  <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M12 8a3 3 0 0 0 3-3 3 3 0 0 0-3-3 3 3 0 0 0-3 3 3 3 0 0 0 3 3m0 3.54C9.64 9.35 6.5 8 3 8v11c3.5 0 6.64 1.35 9 3.54 2.36-2.19 5.5-3.54 9-3.54V8c-3.5 0-6.64 1.35-9 3.54"/></svg>

    </a>
    Plant Spatial Multiomics Analysis
  </label>
  
    <div class="md-nav__source">
      <a href="https://github.com/YAOJ-bioin/Plant_spatial_multiomics_analyis" title="Go to repository" class="md-source" data-md-component="source">
  <div class="md-source__icon md-icon">
    
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 448 512"><!--! Font Awesome Free 6.7.2 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2024 Fonticons, Inc.--><path d="M439.55 236.05 244 40.45a28.87 28.87 0 0 0-40.81 0l-40.66 40.63 51.52 51.52c27.06-9.14 52.68 16.77 43.39 43.68l49.66 49.66c34.23-11.8 61.18 31 35.47 56.69-26.49 26.49-70.21-2.87-56-37.34L240.22 199v121.85c25.3 12.54 22.26 41.85 9.08 55a34.34 34.34 0 0 1-48.55 0c-17.57-17.6-11.07-46.91 11.25-56v-123c-20.8-8.51-24.6-30.74-18.64-45L142.57 101 8.45 235.14a28.86 28.86 0 0 0 0 40.81l195.61 195.6a28.86 28.86 0 0 0 40.8 0l194.69-194.69a28.86 28.86 0 0 0 0-40.81"/></svg>
  </div>
  <div class="md-source__repository">
    YAOJ-bioin/Plant_spatial_multiomics_analyis
  </div>
</a>
    </div>
  
  <ul class="md-nav__list" data-md-scrollfix>
    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href=".." class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Welcome
  </span>
  

      </a>
    </li>
  

    
      
      
  
  
    
  
  
  
    <li class="md-nav__item md-nav__item--active">
      
      <input class="md-nav__toggle md-toggle" type="checkbox" id="__toc">
      
      
        
      
      
        <label class="md-nav__link md-nav__link--active" for="__toc">
          
  
  <span class="md-ellipsis">
    Preprocessing
  </span>
  

          <span class="md-nav__icon md-icon"></span>
        </label>
      
      <a href="./" class="md-nav__link md-nav__link--active">
        
  
  <span class="md-ellipsis">
    Preprocessing
  </span>
  

      </a>
      
        

<nav class="md-nav md-nav--secondary" aria-label="Table of contents">
  
  
  
    
  
  
    <label class="md-nav__title" for="__toc">
      <span class="md-nav__icon md-icon"></span>
      Table of contents
    </label>
    <ul class="md-nav__list" data-md-component="toc" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#1-data-aquisition" class="md-nav__link">
    <span class="md-ellipsis">
      1. Data aquisition
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#2-define-spatial-units" class="md-nav__link">
    <span class="md-ellipsis">
      2. Define spatial units
    </span>
  </a>
  
    <nav class="md-nav" aria-label="2. Define spatial units">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#21-bin-segmentation" class="md-nav__link">
    <span class="md-ellipsis">
      2.1 Bin segmentation
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#22-cell-segmentation" class="md-nav__link">
    <span class="md-ellipsis">
      2.2 Cell segmentation
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#3-construct-a-spatial-matrix-with-coordinates" class="md-nav__link">
    <span class="md-ellipsis">
      3. Construct a spatial matrix with coordinates
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#4-quality-control-and-data-optimization" class="md-nav__link">
    <span class="md-ellipsis">
      4. Quality control and data optimization
    </span>
  </a>
  
    <nav class="md-nav" aria-label="4. Quality control and data optimization">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#41-quality-control" class="md-nav__link">
    <span class="md-ellipsis">
      4.1 Quality control
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#42-data-imputation" class="md-nav__link">
    <span class="md-ellipsis">
      4.2 data imputation
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#43-batch-correction" class="md-nav__link">
    <span class="md-ellipsis">
      4.3 batch correction
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
    </ul>
  
</nav>
      
    </li>
  

    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="../basic_analysis.md" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Basic analysis
  </span>
  

      </a>
    </li>
  

    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="../multiomics_integration.md" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Multiomics integration
  </span>
  

      </a>
    </li>
  

    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="../downstream_analysis.md" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Downstream analysis
  </span>
  

      </a>
    </li>
  

    
  </ul>
</nav>
                  </div>
                </div>
              </div>
            
            
              
              <div class="md-sidebar md-sidebar--secondary" data-md-component="sidebar" data-md-type="toc" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    

<nav class="md-nav md-nav--secondary" aria-label="Table of contents">
  
  
  
    
  
  
    <label class="md-nav__title" for="__toc">
      <span class="md-nav__icon md-icon"></span>
      Table of contents
    </label>
    <ul class="md-nav__list" data-md-component="toc" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#1-data-aquisition" class="md-nav__link">
    <span class="md-ellipsis">
      1. Data aquisition
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#2-define-spatial-units" class="md-nav__link">
    <span class="md-ellipsis">
      2. Define spatial units
    </span>
  </a>
  
    <nav class="md-nav" aria-label="2. Define spatial units">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#21-bin-segmentation" class="md-nav__link">
    <span class="md-ellipsis">
      2.1 Bin segmentation
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#22-cell-segmentation" class="md-nav__link">
    <span class="md-ellipsis">
      2.2 Cell segmentation
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#3-construct-a-spatial-matrix-with-coordinates" class="md-nav__link">
    <span class="md-ellipsis">
      3. Construct a spatial matrix with coordinates
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#4-quality-control-and-data-optimization" class="md-nav__link">
    <span class="md-ellipsis">
      4. Quality control and data optimization
    </span>
  </a>
  
    <nav class="md-nav" aria-label="4. Quality control and data optimization">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#41-quality-control" class="md-nav__link">
    <span class="md-ellipsis">
      4.1 Quality control
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#42-data-imputation" class="md-nav__link">
    <span class="md-ellipsis">
      4.2 data imputation
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#43-batch-correction" class="md-nav__link">
    <span class="md-ellipsis">
      4.3 batch correction
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
    </ul>
  
</nav>
                  </div>
                </div>
              </div>
            
          
          
            <div class="md-content" data-md-component="content">
              <article class="md-content__inner md-typeset">
                
                  

  
  


<h1 id="preprocess">Preprocess</h1>
<h2 id="1-data-aquisition">1. Data aquisition</h2>
<p>We categorize the preprocessing of plant ST data into four parts: data acquisition, defining spatial units, constructing a spatial matrix with coordinates, and quality control. </p>
<p>Here we use the data from the paper <a href="https://doi.org/10.1016/j.devcel.2024.05.016">"Spatiotemporal transcriptomic landscape of rice embryonic cells during seed germination"</a>, it is a <code>rice seed embryo</code> spatial transcriptomics dataset from <code>stereo-seq</code>.
- The data has been aligned, quantified and barcode mapped through the <a href="https://github.com/STOmics/SAW">SAW workflow</a>
- The data input here is GEM file, which is the output of the SAW workflow.
- The dataset is available at Spatial Transcript Omics DataBase, STOmics DB <a href="https://db.cngb.org/stomics/project/STT0000049/spatialGeneExpression">STT0000049</a>
- The pipiline <a href="https://stereopy.readthedocs.io/en/latest/index.html">stereppy</a> is used to process the data. It accepts the GEM/GEF file (from Sterep-seq) and H5ad (output from scanpy) as input.</p>
<div class="language-python highlight"><pre><span></span><code><span id="__span-0-1"><a id="__codelineno-0-1" name="__codelineno-0-1" href="#__codelineno-0-1"></a><span class="kn">import</span><span class="w"> </span><span class="nn">sys</span>
</span><span id="__span-0-2"><a id="__codelineno-0-2" name="__codelineno-0-2" href="#__codelineno-0-2"></a><span class="kn">import</span><span class="w"> </span><span class="nn">os</span>
</span><span id="__span-0-3"><a id="__codelineno-0-3" name="__codelineno-0-3" href="#__codelineno-0-3"></a><span class="kn">import</span><span class="w"> </span><span class="nn">warnings</span>
</span><span id="__span-0-4"><a id="__codelineno-0-4" name="__codelineno-0-4" href="#__codelineno-0-4"></a><span class="kn">import</span><span class="w"> </span><span class="nn">stereo</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">st</span>
</span><span id="__span-0-5"><a id="__codelineno-0-5" name="__codelineno-0-5" href="#__codelineno-0-5"></a><span class="n">warnings</span><span class="o">.</span><span class="n">filterwarnings</span><span class="p">(</span><span class="s1">&#39;ignore&#39;</span><span class="p">)</span>
</span></code></pre></div>
<style>.bk-root, .bk-root .bk:before, .bk-root .bk:after {
  font-family: var(--jp-ui-font-size1);
  font-size: var(--jp-ui-font-size1);
  color: var(--jp-ui-font-color1);
}
</style>

<div class="language-python highlight"><pre><span></span><code><span id="__span-1-1"><a id="__codelineno-1-1" name="__codelineno-1-1" href="#__codelineno-1-1"></a><span class="n">data_path</span> <span class="o">=</span> <span class="s1">&#39;/home/yaoj/Review_plant_Spatial_multiomics_data_analysis/data/A5_1_bin1.gem.gz&#39;</span>
</span><span id="__span-1-2"><a id="__codelineno-1-2" name="__codelineno-1-2" href="#__codelineno-1-2"></a><span class="c1"># read GEM file information</span>
</span><span id="__span-1-3"><a id="__codelineno-1-3" name="__codelineno-1-3" href="#__codelineno-1-3"></a><span class="n">st</span><span class="o">.</span><span class="n">io</span><span class="o">.</span><span class="n">read_gem</span><span class="p">(</span><span class="n">data_path</span><span class="p">)</span>
</span></code></pre></div>
<h2 id="2-define-spatial-units">2. Define spatial units</h2>
<h3 id="21-bin-segmentation">2.1 Bin segmentation</h3>
<p>Bin segmentation is simpler and more widely adopted, where the key step involves determining the appropriate bin size. The average cell size within the tissue is estimated using image information, and tools like <code>ImageJ</code>, and <code>CaseViewer</code> are often employed. The bin size can then be optimized based on clustering results, ensuring the clusters correspond to distinct cell types and are biologically interpretable </p>
<p>We estimate the average cell size in this rice seed embryo dataset is 25 μm².
- Cell side length = X * ST_resolution
- X is the bin size, in Bin 50, stereo-seq resolution is 0.5 um, the cell side length is 50 * 0.5 = 25 um</p>
<p><div class="language-python highlight"><pre><span></span><code><span id="__span-2-1"><a id="__codelineno-2-1" name="__codelineno-2-1" href="#__codelineno-2-1"></a><span class="c1"># define the bin size= 50</span>
</span><span id="__span-2-2"><a id="__codelineno-2-2" name="__codelineno-2-2" href="#__codelineno-2-2"></a><span class="n">data</span> <span class="o">=</span> <span class="n">st</span><span class="o">.</span><span class="n">io</span><span class="o">.</span><span class="n">read_gem</span><span class="p">(</span><span class="n">file_path</span><span class="o">=</span><span class="n">data_path</span><span class="p">,</span> <span class="n">bin_size</span><span class="o">=</span><span class="mi">50</span><span class="p">)</span>
</span><span id="__span-2-3"><a id="__codelineno-2-3" name="__codelineno-2-3" href="#__codelineno-2-3"></a><span class="n">data</span>
</span></code></pre></div>
    StereoExpData object with n_cells X n_genes = 2566 X 25320
    bin_type: bins
    bin_size: 50
    offset_x = 10025
    offset_y = 8025
    cells: ['cell_name']
    genes: ['gene_name']
    cells_matrix = ['spatial']
    Layers with keys: 
    result: []</p>
<div class="language-python highlight"><pre><span></span><code><span id="__span-3-1"><a id="__codelineno-3-1" name="__codelineno-3-1" href="#__codelineno-3-1"></a><span class="c1">#data.tl.cal_qc()</span>
</span><span id="__span-3-2"><a id="__codelineno-3-2" name="__codelineno-3-2" href="#__codelineno-3-2"></a><span class="n">data</span><span class="o">.</span><span class="n">plt</span><span class="o">.</span><span class="n">spatial_scatter</span><span class="p">()</span>
</span></code></pre></div>
<div class="language-text highlight"><pre><span></span><code>[2025-01-18 03:31:51][Stereo][433552][MainThread][140065997267328][st_pipeline][41][INFO]: start to run cal_qc...
[2025-01-18 03:31:52][Stereo][433552][MainThread][140065997267328][st_pipeline][44][INFO]: cal_qc end, consume time 0.1649s.

BokehModel(combine_events=True, render_bundle={&#39;docs_json&#39;: {&#39;eee6ab55-1588-4559-888b-5eee0a19dc22&#39;: {&#39;defs&#39;: …
</code></pre></div>
<p><img alt="png" src="../1_preprocess_files/1_preprocess_5_2.png" /></p>
<p>Test bin size = 30,50,80,100, and compare the results</p>
<div class="language-python highlight"><pre><span></span><code><span id="__span-4-1"><a id="__codelineno-4-1" name="__codelineno-4-1" href="#__codelineno-4-1"></a><span class="kn">import</span><span class="w"> </span><span class="nn">panel</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">pn</span>
</span><span id="__span-4-2"><a id="__codelineno-4-2" name="__codelineno-4-2" href="#__codelineno-4-2"></a><span class="k">def</span><span class="w"> </span><span class="nf">compare_multiple_bins</span><span class="p">(</span><span class="n">file_path</span><span class="p">,</span> <span class="n">bin_sizes</span><span class="o">=</span><span class="p">[</span><span class="mi">30</span><span class="p">,</span> <span class="mi">50</span><span class="p">,</span> <span class="mi">80</span><span class="p">,</span> <span class="mi">100</span><span class="p">]):</span>
</span><span id="__span-4-3"><a id="__codelineno-4-3" name="__codelineno-4-3" href="#__codelineno-4-3"></a>    <span class="n">plots</span> <span class="o">=</span> <span class="p">[]</span>  <span class="c1"># 存储所有图形对象</span>
</span><span id="__span-4-4"><a id="__codelineno-4-4" name="__codelineno-4-4" href="#__codelineno-4-4"></a>    <span class="k">for</span> <span class="n">bin_size</span> <span class="ow">in</span> <span class="n">bin_sizes</span><span class="p">:</span>
</span><span id="__span-4-5"><a id="__codelineno-4-5" name="__codelineno-4-5" href="#__codelineno-4-5"></a>        <span class="c1"># 读取数据</span>
</span><span id="__span-4-6"><a id="__codelineno-4-6" name="__codelineno-4-6" href="#__codelineno-4-6"></a>        <span class="n">data</span> <span class="o">=</span> <span class="n">st</span><span class="o">.</span><span class="n">io</span><span class="o">.</span><span class="n">read_gem</span><span class="p">(</span><span class="n">file_path</span><span class="o">=</span><span class="n">file_path</span><span class="p">,</span> <span class="n">bin_size</span><span class="o">=</span><span class="n">bin_size</span><span class="p">)</span>
</span><span id="__span-4-7"><a id="__codelineno-4-7" name="__codelineno-4-7" href="#__codelineno-4-7"></a>        <span class="n">data</span><span class="o">.</span><span class="n">tl</span><span class="o">.</span><span class="n">cal_qc</span><span class="p">()</span>
</span><span id="__span-4-8"><a id="__codelineno-4-8" name="__codelineno-4-8" href="#__codelineno-4-8"></a>
</span><span id="__span-4-9"><a id="__codelineno-4-9" name="__codelineno-4-9" href="#__codelineno-4-9"></a>        <span class="c1"># plot spatial scatter</span>
</span><span id="__span-4-10"><a id="__codelineno-4-10" name="__codelineno-4-10" href="#__codelineno-4-10"></a>        <span class="n">plot</span> <span class="o">=</span> <span class="n">data</span><span class="o">.</span><span class="n">plt</span><span class="o">.</span><span class="n">spatial_scatter</span><span class="p">(</span>
</span><span id="__span-4-11"><a id="__codelineno-4-11" name="__codelineno-4-11" href="#__codelineno-4-11"></a>            <span class="n">cells_key</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;n_genes_by_counts&#39;</span><span class="p">],</span>
</span><span id="__span-4-12"><a id="__codelineno-4-12" name="__codelineno-4-12" href="#__codelineno-4-12"></a>            <span class="n">width</span><span class="o">=</span><span class="mi">2</span><span class="p">,</span>
</span><span id="__span-4-13"><a id="__codelineno-4-13" name="__codelineno-4-13" href="#__codelineno-4-13"></a>            <span class="n">height</span><span class="o">=</span><span class="mi">2</span><span class="p">,</span>
</span><span id="__span-4-14"><a id="__codelineno-4-14" name="__codelineno-4-14" href="#__codelineno-4-14"></a>            <span class="n">title</span><span class="o">=</span><span class="p">[</span><span class="sa">f</span><span class="s1">&#39;bin size = </span><span class="si">{</span><span class="n">bin_size</span><span class="si">}</span><span class="s1">&#39;</span><span class="p">]</span>
</span><span id="__span-4-15"><a id="__codelineno-4-15" name="__codelineno-4-15" href="#__codelineno-4-15"></a>        <span class="p">)</span>
</span><span id="__span-4-16"><a id="__codelineno-4-16" name="__codelineno-4-16" href="#__codelineno-4-16"></a>        <span class="n">plots</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">plot</span><span class="p">)</span>
</span><span id="__span-4-17"><a id="__codelineno-4-17" name="__codelineno-4-17" href="#__codelineno-4-17"></a>
</span><span id="__span-4-18"><a id="__codelineno-4-18" name="__codelineno-4-18" href="#__codelineno-4-18"></a>    <span class="c1"># use panel&#39;s Row layout to display all plots side by side</span>
</span><span id="__span-4-19"><a id="__codelineno-4-19" name="__codelineno-4-19" href="#__codelineno-4-19"></a>    <span class="k">return</span> <span class="n">pn</span><span class="o">.</span><span class="n">Row</span><span class="p">(</span><span class="o">*</span><span class="n">plots</span><span class="p">)</span>
</span><span id="__span-4-20"><a id="__codelineno-4-20" name="__codelineno-4-20" href="#__codelineno-4-20"></a>
</span><span id="__span-4-21"><a id="__codelineno-4-21" name="__codelineno-4-21" href="#__codelineno-4-21"></a><span class="c1"># use the function</span>
</span><span id="__span-4-22"><a id="__codelineno-4-22" name="__codelineno-4-22" href="#__codelineno-4-22"></a><span class="n">pn</span><span class="o">.</span><span class="n">extension</span><span class="p">()</span>  <span class="c1"># initialize panel</span>
</span><span id="__span-4-23"><a id="__codelineno-4-23" name="__codelineno-4-23" href="#__codelineno-4-23"></a><span class="n">combined_plot</span> <span class="o">=</span> <span class="n">compare_multiple_bins</span><span class="p">(</span><span class="n">data_path</span><span class="p">)</span>
</span><span id="__span-4-24"><a id="__codelineno-4-24" name="__codelineno-4-24" href="#__codelineno-4-24"></a><span class="n">combined_plot</span>
</span></code></pre></div>
<p><img alt="Bin size" src="../assets/images/Binsize.png" /></p>
<h3 id="22-cell-segmentation">2.2 Cell segmentation</h3>
<p><strong>Cell segmentation</strong> offers a more granular approach, especially in tissues where cell size exhibit considerable variability. Recent advancements in segmentation tools allow for more precise delineation of individual cells, typically using cell wall staining images and manually annotated training models. Popular tools include <code>Cellpose2/3</code>, <code>StarDist</code>. Comprehensive ST frameworks, such as <code>STCellBin</code>, support segmentation based on models trained with Arabidopsis seed cell segmentation ground truth.</p>
<ul>
<li>Here we use <code>Cellpose2</code> to segment the cells.</li>
<li>The model we trained according to the cell wall staining images of rice seed embryo is available at <a href="https://github.com/YAOJ-bioin/REGSTA-script">https://github.com/YAOJ-bioin/REGSTA-script</a></li>
</ul>
<p><div class="language-powershell highlight"><pre><span></span><code><span id="__span-5-1"><a id="__codelineno-5-1" name="__codelineno-5-1" href="#__codelineno-5-1"></a><span class="c">##cell segmentation</span>
</span><span id="__span-5-2"><a id="__codelineno-5-2" name="__codelineno-5-2" href="#__codelineno-5-2"></a><span class="c">#cellpose installation</span>
</span><span id="__span-5-3"><a id="__codelineno-5-3" name="__codelineno-5-3" href="#__codelineno-5-3"></a><span class="c">#download environment.yml from https://github.com/MouseLand/cellpose</span>
</span><span id="__span-5-4"><a id="__codelineno-5-4" name="__codelineno-5-4" href="#__codelineno-5-4"></a><span class="c">#cmd</span>
</span><span id="__span-5-5"><a id="__codelineno-5-5" name="__codelineno-5-5" href="#__codelineno-5-5"></a><span class="n">conda</span> <span class="n">env</span> <span class="n">create</span> <span class="o">-f</span> <span class="n">environment</span><span class="p">.</span><span class="n">yml</span>
</span><span id="__span-5-6"><a id="__codelineno-5-6" name="__codelineno-5-6" href="#__codelineno-5-6"></a><span class="n">conda</span> <span class="n">activate</span> <span class="n">cellpose</span>
</span><span id="__span-5-7"><a id="__codelineno-5-7" name="__codelineno-5-7" href="#__codelineno-5-7"></a><span class="n">pip</span> <span class="n">install</span> <span class="n">cellpose</span>
</span><span id="__span-5-8"><a id="__codelineno-5-8" name="__codelineno-5-8" href="#__codelineno-5-8"></a>
</span><span id="__span-5-9"><a id="__codelineno-5-9" name="__codelineno-5-9" href="#__codelineno-5-9"></a>
</span><span id="__span-5-10"><a id="__codelineno-5-10" name="__codelineno-5-10" href="#__codelineno-5-10"></a><span class="c">#prediction:</span>
</span><span id="__span-5-11"><a id="__codelineno-5-11" name="__codelineno-5-11" href="#__codelineno-5-11"></a><span class="n">python</span> <span class="n">-m</span> <span class="n">cellpose</span>  <span class="p">-</span><span class="n">-exclude_on_edges</span>  <span class="p">-</span><span class="n">-dir</span> <span class="p">.\</span><span class="n">data4test</span><span class="p">\</span><span class="n">img</span> <span class="p">-</span><span class="n">-diameter</span> <span class="n">0</span>  <span class="p">-</span><span class="n">-save_outlines</span>  <span class="p">-</span><span class="n">-save_txt</span>  <span class="p">-</span><span class="n">-in_folders</span> <span class="p">-</span><span class="n">-verbose</span> <span class="p">-</span><span class="n">-flow_threshold</span> <span class="n">5</span> <span class="p">-</span><span class="n">-pretrained_model</span> <span class="p">.\</span><span class="n">model</span><span class="p">\</span><span class="n">cellpose_residual_on_style_on_concatenation_off_CP_1_2022_11_26_02_48_20</span><span class="p">.</span><span class="n">469512</span>
</span><span id="__span-5-12"><a id="__codelineno-5-12" name="__codelineno-5-12" href="#__codelineno-5-12"></a>
</span><span id="__span-5-13"><a id="__codelineno-5-13" name="__codelineno-5-13" href="#__codelineno-5-13"></a><span class="c">#--dir  (only slice images saving path)</span>
</span><span id="__span-5-14"><a id="__codelineno-5-14" name="__codelineno-5-14" href="#__codelineno-5-14"></a><span class="c">#--save_txt   (&quot;outlices.txt&quot; whice serves as cell segmentation result in next step ,will be saved in a folder which is named &quot;txt_outlines&quot;)</span>
</span><span id="__span-5-15"><a id="__codelineno-5-15" name="__codelineno-5-15" href="#__codelineno-5-15"></a><span class="c">#--pretrained_model  (path of auto-segmentation model)</span>
</span></code></pre></div>
<img alt="Cellsegmentation" src="../assets/images/cell_segmentation.png" /></p>
<h2 id="3-construct-a-spatial-matrix-with-coordinates">3. Construct a spatial matrix with coordinates</h2>
<h2 id="4-quality-control-and-data-optimization">4. Quality control and data optimization</h2>
<h3 id="41-quality-control">4.1 Quality control</h3>
<h3 id="42-data-imputation">4.2 data imputation</h3>
<h3 id="43-batch-correction">4.3 batch correction</h3>












                
              </article>
            </div>
          
          
<script>var target=document.getElementById(location.hash.slice(1));target&&target.name&&(target.checked=target.name.startsWith("__tabbed_"))</script>
        </div>
        
      </main>
      
        <footer class="md-footer">
  
  <div class="md-footer-meta md-typeset">
    <div class="md-footer-meta__inner md-grid">
      <div class="md-copyright">
  
  
    Made with
    <a href="https://squidfunk.github.io/mkdocs-material/" target="_blank" rel="noopener">
      Material for MkDocs
    </a>
  
</div>
      
    </div>
  </div>
</footer>
      
    </div>
    <div class="md-dialog" data-md-component="dialog">
      <div class="md-dialog__inner md-typeset"></div>
    </div>
    
    
    <script id="__config" type="application/json">{"base": "..", "features": ["navigation.instant"], "search": "../assets/javascripts/workers/search.f8cc74c7.min.js", "translations": {"clipboard.copied": "Copied to clipboard", "clipboard.copy": "Copy to clipboard", "search.result.more.one": "1 more on this page", "search.result.more.other": "# more on this page", "search.result.none": "No matching documents", "search.result.one": "1 matching document", "search.result.other": "# matching documents", "search.result.placeholder": "Type to start searching", "search.result.term.missing": "Missing", "select.version": "Select version"}}</script>
    
    
      <script src="../assets/javascripts/bundle.60a45f97.min.js"></script>
      
    
  </body>
</html>